src/validation.h

```diff
@@ -187,34 +188,65 @@ void PruneBlockFilesManual(CChainState& active_chainstate, int nManualPruneHeigh
 * Validation result for a single transaction mempool acceptance.
 */
 struct MempoolAcceptResult {
-    /** Used to indicate the results of mempool validation,
-    * including the possibility of unfinished validation.
-    */
+    /** Used to indicate the results of mempool validation. */
     enum class ResultType {
         VALID, //!> Fully validated, valid.
         INVALID, //!> Invalid.
     };
-    ResultType m_result_type;
-    TxValidationState m_state;
+    const ResultType m_result_type;
+    const TxValidationState m_state;
 
     // The following fields are only present when m_result_type = ResultType::VALID
     /** Mempool transactions replaced by the tx per BIP 125 rules. */
-    std::optional<std::list<CTransactionRef>> m_replaced_transactions;
-    /** Raw base fees. */
-    std::optional<CAmount> m_base_fees;
+    const std::optional<std::list<CTransactionRef>> m_replaced_transactions;
+    /** Raw base fees in satoshis. */
+    const std::optional<CAmount> m_base_fees;
+
+    static MempoolAcceptResult Failure(const TxValidationState state) {
+        return MempoolAcceptResult(state);
+    }
+
+    static MempoolAcceptResult Success(std::list<CTransactionRef>&& replaced_txns, const CAmount fees) {
+        return MempoolAcceptResult(std::move(replaced_txns), fees);
+    }
 
+// Private constructors. Use static methods MempoolAcceptResult::Success, etc. to construct.
+private:
     /** Constructor for failure case */
-    explicit MempoolAcceptResult(TxValidationState state)
-        : m_result_type(ResultType::INVALID), m_state(state) {
+    explicit MempoolAcceptResult(const TxValidationState state)
+        : m_result_type{ResultType::INVALID}, m_state{state}, m_replaced_transactions(std::nullopt),
+        m_base_fees{std::nullopt} {
             Assume(!state.IsValid()); // Can be invalid or error
         }
 
     /** Constructor for success case */
-    explicit MempoolAcceptResult(std::list<CTransactionRef>&& replaced_txns, CAmount fees)
-        : m_result_type(ResultType::VALID), m_state(TxValidationState{}),
+    explicit MempoolAcceptResult(std::list<CTransactionRef>&& replaced_txns, const CAmount fees)
+        : m_result_type{ResultType::VALID},
         m_replaced_transactions(std::move(replaced_txns)), m_base_fees(fees) {}
 };
 
+/**
+* Validation result for package mempool acceptance.
+*/
+struct PackageMempoolAcceptResult
+{
+    PackageValidationState m_state;
+    /**
+    * Map from txid to finished MempoolAcceptResults. The client is responsible
+    * for keeping track of the transaction objects themselves. If a result is not
+    * present, it means validation was unfinished for that transaction.
+    */
+    std::map<const uint256, const MempoolAcceptResult> m_tx_results;
```

Comment:
Can this not be a vector of equal length to the package sent and index it by index?

This will most likely save memory & less work.

Reply:
I guess that'd need to be a vector of optionals - probably more memory, actually. I only insert a result into the map if it's fully validated, so sometimes it only has 1 result.

Reply:
Well so the max size is a vec of size 25 -- so it's a single allocation when you reserve v.s. 25 allocations for the map.

Further, the map each allocation is IIRC 2-or-3 pointers (16 - 24 bytes), 32 bytes for hash, and then however big MempoolAcceptResult is.

Further in the case where you do have a few things it's extra work/indirection to iterate/lookup.

You don't need a optional because MempoolAcceptResult can use it's inner enum to have an "uninit" state.

The expected case is that all the txns fully validate right? Wouldn't we disconnect peers that send us things that don't validate a lot?

Reply:
I guess to be more concrete -- allocating 25 kinda small things is a vec as single allocation isn't really an issue, so we might as well do it. In the worst case it's more efficient, and in the best case it's not too bad.